test/ha/activator_test.go

```diff
@@ -0,0 +1,147 @@
+)
+
+// The Activator does not have leader election enabled.
+// The test ensures that stopping one of the activator pods doesn't affect user applications.
+// One service is probed during activator restarts and another service is used for testing
+// that we can scale from zero after activator restart.
+func TestActivatorHA(t *testing.T) {
+	clients := e2e.Setup(t)
+
+	if err := waitForDeploymentScale(clients, activatorDeploymentName, haReplicas); err != nil {
+		t.Fatalf("Deployment %s not scaled to %d: %v", activatorDeploymentName, haReplicas, err)
+	}
+
+	// Create first service that we will continually probe during activator restart.
+	names, resources := createPizzaPlanetService(t,
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.MinScaleAnnotationKey:  "1",  //make sure we don't scale to zero during the test
+			autoscaling.TargetBurstCapacityKey: "-1", // make sure all requests go through the activator
+		}),
+	)
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	// Create second service that will be scaled to zero and after stopping the activator we'll
+	// ensure it can be scaled back from zero.
+	namesScaleToZero, resourcesScaleToZero := createPizzaPlanetService(t,
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.WindowAnnotationKey:    autoscaling.WindowMin.String(), //make sure we scale to zero quickly
+			autoscaling.TargetBurstCapacityKey: "-1",                           // make sure all requests go through the activator
+		}),
+	)
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, namesScaleToZero) })
+	defer test.TearDown(clients, namesScaleToZero)
+
+	if err := e2e.WaitForScaleToZero(t, revisionresourcenames.Deployment(resourcesScaleToZero.Revision), clients); err != nil {
+		t.Fatal("Failed to scale to zero:", err)
+	}
+
+	scaleToZeroURL := resources.Service.Status.URL.URL()
+	prober := test.RunRouteProber(log.Printf, clients, resources.Service.Status.URL.URL())
+	defer assertSLO(t, prober)
+
+	spoofingClient, err := pkgTest.NewSpoofingClient(clients.KubeClient, t.Logf, scaleToZeroURL.Hostname(), test.ServingFlags.ResolvableDomain, test.AddRootCAtoTransport(t.Logf, clients, test.ServingFlags.Https))
+	if err != nil {
+		t.Fatal("Error creating spoofing client:", err)
+	}
+
+	pods, err := clients.KubeClient.Kube.CoreV1().Pods(test.ServingFlags.SystemNamespace).List(metav1.ListOptions{
+		LabelSelector: activatorLabel,
+	})
+	if err != nil {
+		t.Fatal("Failed to get activator pods:", err)
+	}
+	activatorPod := pods.Items[0].Name // stop the oldest activator pod
+
+	clients.KubeClient.Kube.CoreV1().Pods(test.ServingFlags.SystemNamespace).Delete(activatorPod, &metav1.DeleteOptions{
+		GracePeriodSeconds: ptr.Int64(0),
+	})
+
+	// Wait for the killed activator to disappear from the knative service's endpoints.
+	if err := waitForChangedPublicEndpoints(t, clients, resourcesScaleToZero.Revision.Name); err != nil {
```

Comment:
There's a race here, where, strictly theoretically, by the time the public endpoints are checked, they are already modified to be the new ones and this will never happen.
I'd query the existing public EPS before and pass that as a param here to ensure there's a change from the previous set.

Reply:
Thanks for the review and suggestions. I know it's taking long. Hopefully with my latest commit we're almost there?